packages/pyright-internal/src/analyzer/semanticTokensWalker.ts

```diff
@@ -548,14 +555,69 @@ export class SemanticTokensWalker extends ParseTreeWalker {
                 modifiers.push(CustomSemanticTokenModifiers.classMember);
 
                 const declaredType = this._evaluator.getTypeForDeclaration(decl)?.type;
-                // the canonical check for properties (used e.g. in the hover message)
-                if (declaredType && isMaybeDescriptorInstance(declaredType)) {
-                    if (declarations.every((d) => this._missingPropertySetter(d))) {
-                        modifiers.push(SemanticTokenModifiers.readonly);
+
+                // if there is no type information or it is not (potentially) a descriptor instance,
+                // use the default `method` token type
+                if (!declaredType || !isMaybeDescriptorInstance(declaredType)) {
+                    return SemanticTokenTypes.method;
+                }
+
+                const isProp = isProperty(declaredType);
+                // if the descriptor instance is a property, check the declarations for `fset` information
+                // if it is not a property, use a check for the presence of `__set__`
+                const isReadOnly = isProp
+                    ? declarations.every((d) => this._missingPropertySetter(d))
+                    : !isMaybeDescriptorInstance(declaredType, true);
+                if (isReadOnly) {
+                    modifiers.push(SemanticTokenModifiers.readonly);
+                }
+
+                const isWrite = isWriteAccess(node);
```

Comment:
is this `isWriteAccess` check needed if we already know whether it's readonly? like can't we just use `!isReadOnly`?

or is one checking whether it's a read access on a property that has a setter? if so maybe add a comment explaining the difference between the two variables